### PR TITLE
Fix issue #142, #159 by updating field parsing logics and templates

### PR DESCRIPTION
**Pull request by ACMoretxj**
* Update `Resource/Patient` template, add repeatable statements for `PID-5` and `PID-11`.
* Update `ADT_A01` template, change `PR1` handling logic from handle once to handle multiple times.
* Update `Hl7v2DataParser`, handle repeatable fields firstly, then parse the whole field, if the field is repeatable, take the first element of `Repeats` as the component source for the whole field.

### CODE REVIEW

**Comment by camila8** [comment]
I merged this branch and now im getting a tag error:
image

**Comment by camila8** [comment]
image
when merging I have found an issue with the provenance. MSH is off by one. It is looking at MSH 8 instead of MSH 7.

**Comment by ACMoretxj** [comment]
> image
> when merging I have found an issue with the provenance. MSH is off by one. It is looking at MSH 8 instead of MSH 7.
> 
> When i put MSH 7 in MSH 8 that error goes away. and then i get a new error:
> image

Can you provide the HL7 message and the main template(such as ADT_A01.liquid) you use? Of course delete the private information before pasting here.

**Comment by camila8** [comment]
Masked PHI:
MSH|^~\&|REG_UPDATE|LFHPL|CLOVE|CLOVE|20200513000000|AKELLER|ADT^A08|616382824|P|2.3|||||||||||
EVN|A08|20200513000000|REG_UPDATE|REG_UPDATE|AKELLER^KELLER^ALAN^JAMES^^^^^NMH^^^^^IPLFN||
PID|1||009107930^^^MR^MR~10012794^^^^NMFF~743620^^^^LFH~000666949734^^^^NMH||_____^______^_||________|M|_____^______^^|_|___ _________ __^^_________^IL^60030-3720^USA^^^____|____|(___)___-____^_^__^^_^___^_______~(___)___-____^_^__^^_^___^_______~^___^________^________@_____.___||___|_||____________|___-__-____|||_||_||||||_||
ZPD||____~____||___ ____|||_|||||||||||||||||_
PD1|||____ ______ ________ __^^_____|_____^_______^_______^^^^^^_____^^^^_____~__________^_______^_______^^^^^^____^^^^___~_____^_______^_______^^^^^^^^^^_____~______^_______^_______^^^^^^^^^^_____~_____^_______^_______^^^^^^______^^^^______||||||||||||||
ROL|_|__|_______|_____^_______^_______^^^^^^_____^^^^_____~__________^_______^_______^^^^^^____^^^^___~_____^_______^_______^^^^^^^^^^_____~______^_______^_______^^^^^^^^^^_____~_____^_______^_______^^^^^^______^^^^______|________||||_______|________|____ _ ________ __^_____ ___^_______^__^_____^___^^^____|(___)___-____^^__^^^___^_______~(___)___-____^^__^^^___^_______
NK1|1|_____^______^^|SP|___ _________ ____^^_________^__^_____^___|(___)___-____^^__^^^___^_______||___|||||||||||||||||||||||||||
NK1|2|||^^^^^___|||___||||||____ ________||||||||||||||||||||____|____
PV1|1|O|________^2415^01^IPLFN^R^^LF^^^^NMHCDEP|ER|||1518282938^GRENNAN^DARA^^^^^^EPIC^^^^NPI~71495^GRENNAN^DARA^^^^^^SE001^^^^SE001~51322^GRENNAN^DARA^^^^^^^^^^NMHMD~951322^GRENNAN^DARA^^^^^^^^^^LFHMD~NM205175^GRENNAN^DARA^^^^^^ADID^^^^ADID|||Hospitalists||||HW|||1518282938^GRENNAN^DARA^^^^^^EPIC^^^^NPI~71495^GRENNAN^DARA^^^^^^SE001^^^^SE001~51322^GRENNAN^DARA^^^^^^^^^^NMHMD~951322^GRENNAN^DARA^^^^^^^^^^LFHMD~NM205175^GRENNAN^DARA^^^^^^ADID^^^^ADID||37768699|H||||||||||||||||||||||||20200512194657|||5401.02|||||||
PV2||___|^_____ _____||||__~__||||||||||||||_|_||||||||||||||||________||_______|||||||||
ZPV|||||^^^^^^^^^^^_-____|||||||______________|______________||________||||||____|____||||||||||||||
ZIF|____-___ ___|
ROL|_|__|__|______^_______^_______^^^^^^______^^^^______~________^_______^_______^^^^^^____^^^^____|______________|||||||
ROL|_|__|__. ________|__________^_______^______^______^^^^^____^^^^___~_____^_______^______^______^^^^^^^^^_____~_____^_______^______^______^^^^^_____^^^^_____~________^_______^______^______^^^^^____^^^^____|______________|______________|||||___ _ _______ __ ___ ___^^_______^__^_____^|(___)___-____^^__^^^___^_______~(___)___-____^^__^^^___^_______
ROL|_|__|_______ ____|______^________^_____^^^^^^______^^^^______~________^________^_____^^^^^^____^^^^____|______________|||||||
ROL|_|__|__|______^_______^____^__^^^^^______^^^^______~________^_______^____^__^^^^^____^^^^____|______________|______________||||||
OBX|1|NM|HT^HEIGHT||68.23|in||||||||20200512||||||||||||||||||
OBX|2|NM|WT^WEIGHT||78|kg||||||||20200512||||||||||||||||||
OBX|3|TX|PRIMARY^INTERFACE PRIMARY CONTACT|1|____________|||||||||20200512||||||||||||||||||
OBX|4|TX|SPECIALFLAG^CONTACT TYPE|1||||||||||20200512||||||||||||||||||
OBX|5|TX|PCLASS^ADT: PATIENT CLASS|1|B|||||||||20200512||||||||||||||||||
OBX|6|NM|AUTHNOTE^CONTACT NUMBER|1||||||||||20200512||||||||||||||||||
OBX|7|TX|ADTPATSTATUS^ADT: PATIENT STATUS|1|2|||||||||20200512||||||||||||||||||
AL1|_|____ _____~____|_____^__ _____ _________^||||||
DG1|_|___-__-__|___._^_________, ___________ ________^___-__-__|_________, ___________ ________||^_____;___||||||||||||||||||||
DG1|_|___-__-__|___.__^_____ _______^___-__-__|_____ _______||^_____;___||||||||||||||||||||
PR1|_|____|_____^_____ _________ _____^____|_____ _________ _____|______________||||||||||
PR1|_|____|_______^_____________^____|_____________|______________||||||||||
PR1|_|____|_______^_______ (______ ____)(___,___,___,___)^____|_______ (______ ____)(___,___,___,___)|______________||||||||||
PR1|_|____|_____^_______ ________ _____^____|_______ ________ _____|______________||||||||||
PR1|_|____|______^___ ___ ____________^____|___ ___ ____________|______________||||||||||
PR1|_|____|______^_______: _____^____|_______: _____|______________||||||||||
PR1|_|____|______^_______: _____^____|_______: _____|______________||||||||||
PR1|_|____|_______^__ __ _/___________ ___ ____^____|__ __ _/___________ ___ ____|______________||||||||||
PR1|_|____|______^_________ _____^____|_________ _____|______________||||||||||
PR1|__|____|_________^.___^____|.___|______________||||||||||
PR1|__|____|_________^.____ ____^____|.____ ____|______________||||||||||
PR1|__|____|________^_____-__ - _____ ____^____|_____-__ - _____ ____|______________||||||||||
PR1|__|____|____^_______ ___________^____|_______ ___________|______________||||||||||
PR1|__|____|_____^__ _____ __ ___^____|__ _____ __ ___|______________||||||||||
PR1|__|____|_______^_____ ____ ___ ____^____|_____ ____ ___ ____|______________||||||||||
PR1|__|____|_______^_____ ________ ___ ____^____|_____ ________ ___ ____|______________||||||||||
PR1|__|____|_______^_____ ___ ____ ____^____|_____ ___ ____ ____|______________||||||||||
PR1|__|____|_____^___ _______ _____ _____^____|___ _______ _____ _____|______________||||||||||
PR1|__|____|_______^___ _____ __^____|___ _____ __|______________||||||||||
PR1|__|____|______^__________ _______ (___) _____^____|__________ _______ (___) _____|______________||||||||||
PR1|__|____|________^___________ ________ _____ (___)^____|___________ ________ _____ (___)|______________||||||||||
PR1|__|____|____^_______ ___________^____|_______ ___________|______________||||||||||
PR1|__|____|______^__________ _______ (___) _____^____|__________ _______ (___) _____|______________||||||||||
PR1|__|____|____^_________________^____|_________________|______________||||||||||
PR1|__|____|________^______ ___________ ____________ _______ ___, _____^____|______ ___________ ____________ _______ ___, _____|______________||||||||||
PR1|__|____|________^______ ___________ ____________ _______ ___, _____^____|______ ___________ ____________ _______ ___, _____|______________||||||||||
PR1|__|____|____^_____ __ ___________ ______^____|_____ __ ___________ ______|______________||||||||||
PR1|__|____|_____^__________ ___^____|__________ ___|______________||||||||||
PR1|__|____|_______^_____ _____ \T\ ______ __ _________^____|_____ _____ \T\ ______ __ _________|______________||||||||||
PR1|__|____|________^_______ _____ _________^____|_______ _____ _________|______________||||||||||
PR1|__|____|____^____ ____^____|____ ____: _________ ____ _______|______________||||||||||
PR1|__|____|_______^____^____|____|______________||||||||||
PR1|__|____|______^_____ __________ ___________ ______^____|_____ __________ ___________ ______|______________||||||||||
PR1|__|____|_______^___ _______/________,______ ____________^____|___ _______/________,______ ____________|______________||||||||||
PR1|__|____|______^_________ __________ _______^____|_________ __________ _______|______________||||||||||
PR1|__|____|_____^__________ ___^____|__________ ___|______________||||||||||
PR1|__|____|____^_________________^____|_________________|______________||||||||||
GT1|_|_______|_____^______^_^||___ _________ __^^_________^__^_____-____^___^^^____|(___)___-____^^^^^___^_______||________|_|___|___|___-__-____||||____ ________|^^^^^___|||__|||||||||||||||||||||||||||||
IN1|1|603405^FIRST HEALTH LIMITED BENEFITS GENERIC|6034|COVENTRY HEALTH CARE||||||||________|||___|_____^______^^|______|________|___ _________ __^^_________^IL^_____-____^USA^^^____|||_**_*_|||___|______________|||||||||_______|_________||||||____|_|^^^^^___|||____||
IN2|||||________||||||||||||||||||||||||||||||||||||||||||||||||||||||||_________||(___)___-____^^^^^___^_______|||||||
UB2|||||||_____ __ ___^________|


{
    "resourceType": "Bundle",
    "type": "transaction",
    "entry": [
        {% assign firstSegments = hl7v2Data | get_first_segments: 'PID|PD1|PV1|PV2|AVR|MSH' -%}

        {% evaluate messageHeaderId using 'ID/MessageHeader' MSH: firstSegments.MSH -%}
        {% if messageHeaderId -%}
            {% include 'Resource/MessageHeader' MSH: firstSegments.MSH, ID: messageHeaderID -%}
        {% endif -%}
        
        {% evaluate patientId using 'ID/Patient' PID: firstSegments.PID, type: 'First' -%}
        {% if patientId -%}
            {% assign fullPatientId = patientId | prepend: 'Patient/' -%}
            {% include 'Resource/Patient' PID: firstSegments.PID, PD1: firstSegments.PD1, ID: patientId -%}
        {% endif -%}

        {% evaluate provenanceId using 'ID/Provenance' MSH: firstSegments.MSH, baseID: patientId -%}
        {% if provenanceId -%}
            {% include 'Resource/Provenance' MSH: firstSegments.MSH, ORC: firstSegments.ORC, ID: provenanceId -%}
        {% endif -%}

        {% evaluate accountId using 'ID/Account' CX: firstSegments.PID.3 -%}
        {% if accountId -%}
            {% include 'Resource/Account' PID: firstSegments.PID, ID: accountId -%}
        {% endif -%}

        {% evaluate organizationId using 'ID/Organization' HD: firstSegments.MSH.6 -%}
        {% if organizationId -%}
            {% include 'Resource/Organization' MSH: firstSegments.MSH, ID: organizationId -%}
        {% endif -%}

        {% evaluate encounterId using 'ID/Encounter' PV1: firstSegments.PV1, baseId: patientId -%}
        {% if encounterId -%}
            {% include 'Resource/Encounter' PV1: firstSegments.PV1, PV2: firstSegments.PV2, PID: firstSegments.PID, ID: encounterId -%}

            {% evaluate locationId using 'ID/Location' PL: firstSegments.PV1.3 -%}
            {% if locationId -%}
                {% include 'Resource/Location' PL: firstSegments.PV1.3, ID: locationId -%}
            {% endif -%}

            {% evaluate locationId using 'ID/Location' PL: firstSegments.PV1.6 -%}
            {% if locationId -%}
                {% include 'Resource/Location' PL: firstSegments.PV1.6, ID: locationId -%}
            {% endif -%}

            {% if patientId -%}
                {% include 'Reference/Encounter/Subject' ID: encounterId, REF: fullPatientId -%}
            {% endif -%}
        {% endif -%}

        {% assign pr1SegmentLists = hl7v2Data | get_segment_lists: 'PR1' -%}
        {% for pr1Segment in pr1SegmentLists.PR1 -%}
            {% evaluate procedureId using 'ID/Procedure' PR1: pr1Segment, baseId: patientId -%}
            {% if procedureId -%}
                {% include 'Resource/Procedure' PR1: pr1Segment, ID: procedureId -%}

                {% evaluate locationId using 'ID/Location' PL: pr1Segment.23 -%}
                {% if locationId -%}
                    {% include 'Resource/Location' PL: pr1Segment.23, ID: locationId -%}
                {% endif -%}

                {% if patientId -%}
                    {% include 'Reference/Procedure/Subject' ID: procedureId, REF: fullPatientId -%}
                {% endif -%}
            {% endif -%}
        {% endfor -%}

        {% assign nk1SegmentLists = hl7v2Data | get_segment_lists: 'NK1' -%}
        {% for nk1Segment in nk1SegmentLists.NK1 -%}
            {% include 'Resource/Patient' NK1: nk1Segment, ID: patientId -%}

            {% evaluate relatedPersonId using 'ID/RelatedPerson' NK1: nk1Segment, baseId: patientId -%}
            {% if relatedPersonId -%}
                {% include 'Resource/RelatedPerson' NK1: nk1Segment, ID: relatedPersonId -%}
                {% if patientId -%}
                    {% include 'Reference/RelatedPerson/Patient' ID: relatedPersonId, REF: fullPatientId -%}
                {% endif -%}
            {% endif -%}
        {% endfor -%}
{% comment %} OBX in our ADT messages are not mostly observations
        {% assign obxSegmentLists = hl7v2Data | get_segment_lists: 'OBX' -%}
        {% for obxSegment in obxSegmentLists.OBX -%}
            {% evaluate observationId using 'ID/Observation' OBX: obxSegment, baseId: patientId -%}
            {% if observationId -%}
                {% include 'Resource/Observation' OBX: obxSegment, ID: observationId -%}
                {% if patientId -%}
                    {% include 'Reference/Observation/Subject' ID: observationId, REF: fullPatientId -%}
                {% endif -%}
            {% endif -%}
        {% endfor -%}
{% endcomment %}
        {% assign al1SegmentLists = hl7v2Data | get_segment_lists: 'AL1' -%}
        {% for al1Segment in al1SegmentLists.AL1 -%}
            {% evaluate allergyIntoleranceId using 'ID/AllergyIntolerance' AL1: al1Segment, baseId: patientId -%}
            {% if allergyIntoleranceId -%}
                {% include 'Resource/AllergyIntolerance' AL1: al1Segment, ID: allergyIntoleranceId -%}
                {% if patientId -%}
                    {% include 'Reference/AllergyIntolerance/Patient' ID: allergyIntoleranceId, REF: fullPatientId -%}
                {% endif -%}
            {% endif -%}
        {% endfor -%}

        {% assign dg1SegmentLists = hl7v2Data | get_segment_lists: 'DG1' -%}
        {% for dg1Segment in dg1SegmentLists.DG1 -%}
            {% evaluate conditionId using 'ID/Condition' DG1: dg1Segment, baseId: patientId -%}
            {% if conditionId -%}
                {% assign fullConditionId = conditionId | prepend: 'Condition/' -%}
                {% include 'Resource/Condition' DG1: dg1Segment, ID: conditionId -%}
                {% if patientId -%}
                    {% include 'Reference/Condition/Subject' ID: conditionId, REF: fullPatientId -%}
                {% endif -%}
                {% if encounterId -%}
                    {% include 'Reference/Encounter/Diagnosis_Condition' ID: encounterId, REF: fullConditionId -%}
                {% endif -%}
            {% endif -%}
        {% endfor -%}
    ] 
}

**Comment by camila8** [comment]
copy and paste is really ugly. If you send me your email I can send it to you.

**Comment by ACMoretxj** [comment]
> copy and paste is really ugly. If you send me your email I can send it to you. I did a pull from main and nothing was changed according to git.

Thanks for your feedback, we've discussed in the mail. The datetime error will be solve after upgrading the vs code extension, and the JSON error will be solved in this PR: #168

File: src/Microsoft.Health.Fhir.Liquid.Converter/Hl7v2/Hl7v2DataParser.cs [hunk 1]
```diff
@@ -81,8 +81,14 @@ private List<Hl7v2Field> ParseFields(string dataString, Hl7v2EncodingCharacters
                 {
                     if (!string.IsNullOrEmpty(fieldValues[f]))
                     {
-                        var components = ParseComponents(fieldValues[f], encodingCharacters);
-                        var field = new Hl7v2Field(fieldValues[f], components);
+                        /**
+                         * Take care of repeatable fields firstly, then render the whole field.
+                         * 1. If the config doesn't point out the field actually contains $RepetitionSeparator as repeatable,
```
Comment (BoyaWu10) [inline src/Microsoft.Health.Fhir.Liquid.Converter/Hl7v2/Hl7v2DataParser.cs:86]: What does `config` refer to here?

File: data/Templates/Hl7v2/Resource/_Patient.liquid [hunk 2]
```diff
@@ -53,9 +55,11 @@
         "gender":"{{ PID.8.Value | get_property: 'CodeSystem/Gender', 'code' }}",
         "address":
         [
+            {% for p in PID.11.Repeats -%}
             {
-                {% include 'DataType/XAD' XAD: PID.11 -%}
+                {% include 'DataType/XAD' XAD: p -%}
```
Comment (BoyaWu10) [inline data/Templates/Hl7v2/Resource/_Patient.liquid:60]: We need to make same changes in template generation tool.

Comment (ACMoretxj) [reply data/Templates/Hl7v2/Resource/_Patient.liquid:60]: Yes, this feature has been added in the generation tool.

File: src/Microsoft.Health.Fhir.Liquid.Converter/Hl7v2/Hl7v2DataParser.cs (hunk 1 above)
Comment (ACMoretxj) [reply src/Microsoft.Health.Fhir.Liquid.Converter/Hl7v2/Hl7v2DataParser.cs:86]: The `encodingCharacters`.

File: data/Templates/Hl7v2/Resource/_Patient.liquid (hunk 2 above)
Comment (ACMoretxj) [reply data/Templates/Hl7v2/Resource/_Patient.liquid:60]: By offline discussion, I'll submit the pull request in the generation tool.

File: src/Microsoft.Health.Fhir.Liquid.Converter/Hl7v2/Hl7v2DataParser.cs (hunk 1 above)
Comment (BoyaWu10) [reply src/Microsoft.Health.Fhir.Liquid.Converter/Hl7v2/Hl7v2DataParser.cs:86]: Got it. Let's update the comment here to be more specific.

**Review by BoyaWu10 — APPROVED** [review]

**Pull request merged**